superset-frontend/packages/superset-ui-core/src/color/CategoricalColorScale.ts

```diff
@@ -96,7 +98,22 @@ class CategoricalColorScale extends ExtensibleFunction {
     const newColor = this.scale(cleanedValue);
     if (!color) {
       color = newColor;
+      // make sure we don't overwrite the origin colors
+      if (!isFeatureEnabled(FeatureFlag.USE_ANALAGOUS_COLORS)) {
+        const updatedRange = [...this.originColors];
+        // remove the color option from shared color
+        sharedColorMap.forEach((value, key) => {
+          if (key !== cleanedValue) {
+            const index = updatedRange.indexOf(value);
+            updatedRange.splice(index, 1);
```

Comment:
It looks like this may be removing items from a list. Can you use a filter function here instead?